gobblin-service/src/test/java/org/apache/gobblin/service/modules/flowgraph/BaseFlowGraphTest.java

```diff
@@ -0,0 +1,163 @@
+  private DataNode node3;
+  private FlowEdge edge1;
+  private FlowEdge edge2;
+  private FlowEdge edge3;
+
+  BaseFlowGraph graph;
+  @BeforeClass
+  public void setUp() throws URISyntaxException, ReflectiveOperationException, JobTemplate.TemplateException, SpecNotFoundException,
+             IOException {
+    Properties properties = new Properties();
+    properties.put("key1", "val1");
+    node1 = new BaseDataNode("node1", properties);
+
+    properties = new Properties();
+    properties.put("key2", "val2");
+    node2 = new BaseDataNode("node2", properties);
+
+    properties = new Properties();
+    properties.put("key3", "val3");
+    node3 = new BaseDataNode("node3", properties);
+
+    FlowTemplate flowTemplate1 = new StaticFlowTemplate(new URI("uri1"),"","", ConfigFactory.empty(),null, null, null);
+    FlowTemplate flowTemplate2 = new StaticFlowTemplate(new URI("uri2"),"","", ConfigFactory.empty(),null, null, null);
+    FlowTemplate flowTemplate3 = new StaticFlowTemplate(new URI("uri3"),"","", ConfigFactory.empty(),null, null, null);
+
+    //Create edge instances
+    edge1 = new BaseFlowEdge(Lists.newArrayList("node1", "node2"), flowTemplate1, null, ConfigFactory.empty(), true);
+    edge2 = new BaseFlowEdge(Lists.newArrayList("node2", "node3"), flowTemplate2, null, ConfigFactory.empty(), true);
+    edge3 = new BaseFlowEdge(Lists.newArrayList("node3", "node1"), flowTemplate3, null, ConfigFactory.empty(), true);
+
+    //Create a FlowGraph
+    graph = new BaseFlowGraph();
+
+    //Add nodes
+    graph.addDataNode(node1);
+    graph.addDataNode(node2);
+    graph.addDataNode(node3);
+
+    //Add edges
+    graph.addFlowEdge(edge1);
+    graph.addFlowEdge(edge2);
+    graph.addFlowEdge(edge3);
+  }
+
+  @Test
+  public void testAddDataNode() throws Exception {
+    Assert.assertTrue(graph.getNodes().contains(node1));
+    Assert.assertTrue(graph.getNodes().contains(node2));
+    Assert.assertTrue(graph.getNodes().contains(node3));
+
+    Field field = BaseFlowGraph.class.getDeclaredField("dataNodeMap");
+    field.setAccessible(true);
+    Map<String, DataNode> dataNodeMap = (Map<String, DataNode>) field.get(graph);
+    Assert.assertEquals(dataNodeMap.get("node1"), node1);
+    Assert.assertEquals(dataNodeMap.get("node2"), node2);
+    Assert.assertEquals(dataNodeMap.get("node3"), node3);
+  }
+
+  @Test
+  public void testAddFlowEdge() throws Exception {
+    Assert.assertEquals(((BaseDataNode)graph.getNode("node1")).getFlowEdges().size(), 1);
+    Assert.assertEquals(((BaseDataNode)graph.getNode("node2")).getFlowEdges().size(), 1);
+    Assert.assertEquals(((BaseDataNode)graph.getNode("node3")).getFlowEdges().size(), 1);
+
+    Assert.assertTrue(((BaseDataNode)graph.getNode("node1")).getFlowEdges().contains(edge1));
+    Assert.assertTrue(((BaseDataNode)graph.getNode("node2")).getFlowEdges().contains(edge2));
+    Assert.assertTrue(((BaseDataNode)graph.getNode("node3")).getFlowEdges().contains(edge3));
+  }
+
+  @Test
```

Comment:
Add test dependency on testAddFlowEdge(), else this can execute before the testAddFlowEdge when parallelism is enabled, leading to flaky tests.

Reply:
Done.